apphub/NLP/neural_machine_translation/transformer.ipynb

```diff
@@ -0,0 +1,709 @@
+     ]
+    }
+   ],
+   "source": [
+    "data = pipeline.get_results()\n",
+    "print(\"source after processing:\")\n",
+    "print(data[\"source\"])\n",
+    "print(\"source batch shape:\")\n",
+    "print(data[\"source\"].shape)\n",
+    "print(\"---------------------------------------------------\")\n",
+    "print(\"target after processing:\")\n",
+    "print(data[\"target\"])\n",
+    "print(\"target batch shape:\")\n",
+    "print(data[\"target\"].shape)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "purple-november",
+   "metadata": {},
+   "source": [
+    "# Transformer Architecture\n",
+    "## Attention Unit\n",
+    "The basic form of attention unit is defined in `scaled_dot_product_attention`, given a set of queries(Q), keys(K), and values(Q), it first performs the matrix multiplication of Q and K. The output of the multiplication means the matching score between various elements of Q and K. Then all the weights are normalized across the Keys dimension. Finally, the normalized score will be multiplied by the V to get the final result.  The intuition behind the attention unit is essentially a dictionary look-up with interpolation."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 6,
+   "id": "frank-ontario",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import tensorflow as tf\n",
+    "\n",
+    "def scaled_dot_product_attention(q, k, v, mask):\n",
+    "    matmul_qk = tf.matmul(q, k, transpose_b=True)\n",
+    "    dk = tf.cast(tf.shape(k)[-1], tf.float32)\n",
+    "    scaled_attention_logits = matmul_qk / tf.math.sqrt(dk)\n",
+    "    scaled_attention_logits += (mask * -1e9)  # this is to make the softmax of masked cells to be 0\n",
+    "    attention_weights = tf.nn.softmax(scaled_attention_logits, axis=-1)\n",
+    "    output = tf.matmul(attention_weights, v)\n",
+    "    return output\n",
+    "\n",
+    "\n",
+    "def point_wise_feed_forward_network(em_dim, dff):\n",
+    "    return tf.keras.Sequential([\n",
+    "        tf.keras.layers.Dense(dff, activation='relu'),  # (batch_size, seq_len, dff)\n",
+    "        tf.keras.layers.Dense(em_dim)  # (batch_size, seq_len, em_dim)\n",
+    "    ])\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "general-liberal",
+   "metadata": {},
+   "source": [
+    "## Multi-head Attention\n",
+    "\n",
+    "There are two drawbacks of the attention unit above:\n",
+    "1. The complexity of matrix multiplication is Q(N^3), when batch size or embedding dimension increases, the computation will not scale well. \n",
+    "2. A single attention is limited in expressing local correlation between two words, because it calculates correlation based on normalizing all embeddings dimensions. However, sometimes the overall normalization will remove some interesting local pattern.  A good analogy is to consider single attention unit as globally averaging of a signal whereas moving average is preferred to preserve certain information.\n",
```

Comment:
```suggestion
    "2. A single attention head is limited in expressing local correlation between two words, because it calculates correlation by normalizing all embeddings dimensions. Sometimes this overall normalization will remove interesting local patterns.  A good analogy is to consider a single attention unit as globally averaging a signal whereas a moving average is preferred to preserve certain information.\n",
```